lib/services/keyVault/lib/keyVault.js

```diff
@@ -16,366 +16,677 @@
+  var segments = baseUri.pathname.split('/');
+  if (segments.length !== 3 && segments.length !== 4) {
+    throw new Error(util.format('Invalid %s identifier: %s. Bad number of segments: %d', collection, identifier, segments.length));
+  }
+
+  if (collection !== segments[1]) {
+    throw new Error(util.format('Invalid %s identifier: %s. segment [1] should be "%s", found "%s"', collection, identifier, collection, segments[1]));
+  }
+
+  var vault = util.format('%s//%s', baseUri.protocol, baseUri.host);
+  var name = segments[2];
+  var version = segments.length === 4 ? segments[3] : null;
+  return new ObjectIdentifier(collection, vault, name, version);
+}
+
+/** Parses a string containing a key identifier and returns the ObjectIdentifier object.
+ @param {string} identifier The key identifier (an URI).
+ @return {ObjectIdentifier} An object that represents the key identifier.
+*/
+function parseKeyIdentifier(identifier) {
+  return parseObjectIdentifier('keys', identifier);
+}
+
+exports.parseKeyIdentifier = parseKeyIdentifier;
+
+/** Parses a string containing a secret identifier and returns the ObjectIdentifier object.
+ @param {string} identifier The secret identifier (an URI).
+ @return {ObjectIdentifier} An object that represents the secret identifier.
+*/
+function parseSecretIdentifier(identifier) {
+  return parseObjectIdentifier('secrets', identifier);
+}
+
+exports.parseSecretIdentifier = parseSecretIdentifier;
+
+function createKeyOpRequest(algorithm, data) {
+  var request = {};
+  request.alg = algorithm;
+  request.value = bufferToBase64Url(data);
+  return request;
+}
+
+function createVerifyOpRequest(algorithm, digest, signature) {
+    var request = {};
+    request.alg = algorithm;
+    request.digest = bufferToBase64Url(digest);
+    request.value = bufferToBase64Url(signature);
+    return request;
+}
+
+function createRestoreKeyRequest(backupValue) {
+    var request = {};
+    request.value = bufferToBase64Url(backupValue);
+    return request;
+}
+
+function serializeImportKeyRequest(request) {
+  var result = { rawJsonRequest: JSON.stringify(request) };
+  if (request.key || request.attributes) {
+    result = JSON.parse(result.rawJsonRequest);
+    if (request.key) {
+      jwkBuffersToStrings(request.key, result.key);
+    }
+    if (request.attributes) {
+      attrDatesToNumbers(request.attributes, result.attributes);
+    }
+    result = { rawJsonRequest: JSON.stringify(result) };
+  }
+  return result;
+}
+
+function serializeKeyRequest(request) {
+  var result = { rawJsonRequest: JSON.stringify(request) };
+  if (request.attributes) {
+    result = JSON.parse(result.rawJsonRequest);
+    attrDatesToNumbers(request.attributes, result.attributes);
+    result = { rawJsonRequest: JSON.stringify(result) };
+  }
+  return result;
+}
+
+function valueResponseDeserializer(callback) {
+  return keyResponseDeserializer(function(err, result, response) {
+    if (err || !result || !result.value) {
```

Comment:
this is a little questionable to me. are you return on a error case? If yes, then it is possible the "err" is null, and the downstream code might take a a success case. Otherwise, this code is good.